### ISSUE TITLE
Add 'time' order to tree.nodes()

### ISSUE THREAD

**Issue by jeromekelleher** [tskit-dev/tskit#246]
It is sometimes useful to iterate over nodes in the tree in 'time' order, where we guarantee that the time of each node visited is >= the time of the previous one. Add this to the options in the ``order`` argument.

**Comment by jeromekelleher** [comment]
This is trivial to do:
```python
nodes = sorted([u for u in tree.nodes()], key=lambda u: tree.time(u))
```
the trick is to do it efficiently.

**Comment by benjeffery** [comment]
Thought I'd spend an evening looking at this to see what it's like working with the codebase and as it had some friendly labels! I've implemented your trivial implementation above and an additional one using a priority queue from the `heapq` module. Code at https://github.com/benjeffery/tskit/commit/5feba1b4db5574436613e091b54470765f5e741d

To make it a fairer comparison between the methods, the order of traversal is the reverse of the one specified as this lets the priority queue method start at the root. Maybe there is a use case I am unaware of that strongly motivates the original order, but (unless `sample_lists=True`) you'd have to first traverse to get the samples then work backwards.

I generated some trees to traverse with `msprime.simulate(n)` and then traversed over the first 20 trees. 
![time](https://user-images.githubusercontent.com/8552/67529924-7beaa700-f6b5-11e9-87c3-808917297207.png)
Code for the plot at https://gist.github.com/benjeffery/8c06f75effdc020b09a7ae0e2ab99246
So the priority queue method is pretty slow - this is due to it not being quite the right algorithm as we are not taking advantage of the fact that we only have need for a monotonic priority queue. In this case to doing something like a radix heap (e.g. https://github.com/iwiwi/radix-heap) would be better, but I couldn't find anything to quickly drop into python.

But do we need to? The trivial method is faster than some of the existing methods, about a ~70-80% perf hit compared to `levelorder` and uses `num_nodes * 2 * 8bytes` of additional memory to do the sort.

Anyway, that's how far I got for now, happy to document and then make a PR if needed.

**Comment by jeromekelleher** [comment]
This is great, thanks @benjeffery! We should definitely work this up as a PR.

I forget what the original motivation for having this ordering is, but I think it was from an application where we needed the nodes time-increasing order. Think of it as a stronger form of postorder. I can't really see what one would use time descending down the tree for (but, perhaps we should consider the names so that both are possible?).

Performance-wise, you're right, it looks like there's no point in doing much beyond the trivial method since it's already so fast. But, this is fun, and you've gone to the trouble of putting the harness together so let's explore!

Going leaf-up is possible doing something like this:
```python
heap = []
for u in tree.tree_sequence.samples():
    heapq.heappush((u, tree.time(u))
# Basically same as your algorithm from here
```

Would this be any faster? There's a subtle difference here though, that we'd only visit nodes that are reachable from samples, and so in some weirdo corner cases this traversal would not reach the same set of nodes as other top-down ones.

I'm also amazed that preorder is slower than level order: preorder should be super-simple and fast. I wonder if using the low-level  operations directly would make this quicker (see [here](https://tskit.readthedocs.io/en/latest/data-model.html#tree-structure) for documentation on the underlying tree structures). E.g.,
```python
    def _preorder_traversal(self, node):
        stack = [node]
        while len(stack) > 0:
            u = stack.pop()
            yield u
            v = self.right_child(u)
            while v != NULL:
                  stack.append(v)
                  v = self.left_sib(v)
```

**Comment by benjeffery** [comment]
I've looked a little into the performance of `preorder` vs `levelorder`. I'll checkout the "leaves up" idea another day.
Here's the current time taken for both (100 trees, n=50000):
```
preorder:	19.56s
levelorder:	17.43s
```
Both have an unneeded call to `is_internal` before they add their children to the queue/stack. (`is_internal` calls `is_leaf` which just does `len(self.children(u)) == 0`)
Removing this saves 4 python function calls and a call to `Tree_get_children` in both methods:
```
preorder:	14.20s
levelorder:	9.86s
```
`levelorder` is using a `deque`, changing `preorder` to use one too:
```
preorder:	11.09s
levelorder:	9.75s (change is just noise)
```
At this point the only difference between the methods is `pop` vs `popleft` and an extra call to `reversed` for `preorder`. Removing this call makes them take the same time (also reassuringly breaks the tests!) so `levelorder` still reigns supreme!

I then removed all `.` usage in the tight loop:
```
preorder:	9.54s
levelorder:	8.05s
```
I then tried your more low-level code, with the same tweaks, but the best I could get was:
```
preorder:	11.69s
```
Which I think makes sense as there's quite few more python function calls per loop.

If these functions are really perf sensitive they could be pushed to the C layer, as you can return an iterator from C. I also haven't looked at optimising any C code.

Here's the code of my final fastest versions:
```
    def _preorder_traversal(self, u):
       stack = collections.deque([u])
       #For perf we store these to avoid lookups
       pop = stack.pop
       ext = stack.extend
       children = self.children
       while stack:
           v = pop()
           ext(reversed(children(v)))
           yield v

    def _levelorder_traversal(self, u):
        queue = collections.deque([u])
        #For perf we store these to avoid lookups
        pop = queue.popleft
        ext = queue.extend
        children = self.children
        while queue:
            v = pop()
            ext(children(v))
            yield v
```
If I find time I'll give the other traversal methods the same treatment and do a separate PR.

**Comment by jeromekelleher** [comment]
Very cool, thanks @benjeffery! We should try to make these methods as fast as we reasonably can, since they are basic building blocks.

It might be worth implementing preorder in C as it's the default if there's a significant perf gain, but it's probably pretty marginal --- as soon as anyone starts *doing* anything with the nodes we're generating then the client side processing will surely dominate the time we're spending.

**Comment by jeromekelleher** [comment]
Closed in #399 and #400
